File: pitch-detection/misc/yin/parabolic_interpolation.go
package yin

func ParabolicInterpolation(samples *[]float64, x_ float64) float64 {
	if x_ < 0 {
		return x_
	}
	x := int(x_)
	xAdjusted := 0

	if x < 1 {
		if (*samples)[x] <= (*samples)[x+1] {
			xAdjusted = x
		} else {
			xAdjusted = x + 1
		}
	} else if x > len(*samples)-1 {
		if (*samples)[x] <= (*samples)[x-1] {
			xAdjusted = x
		} else {
			xAdjusted = x - 1
		}
	} else {
		den := (*samples)[x+1] + (*samples)[x-1] - 2*(*samples)[x]
		delta := (*samples)[x-1] - (*samples)[x+1]
		if den == 0.0 {
			return x_
		} else {
			return x_ + delta/(2*den)
		}
	}

	return float64(xAdjusted)
}

File: pitch-detection/src/parabolic_interpolation.cpp
#include "pitch_detection.h"
#include <vector>

template <typename T>
std::pair<T, T>
util::parabolic_interpolation(const std::vector<T> &array, int x_)
{
	int x_adjusted;
	T x = (T)x_;

	if (x < 1) {
		x_adjusted = (array[x] <= array[x + 1]) ? x : x + 1;
	} else if (x > signed(array.size()) - 1) {
		x_adjusted = (array[x] <= array[x - 1]) ? x : x - 1;
	} else {
		T den = array[x + 1] + array[x - 1] - 2 * array[x];
		T delta = array[x - 1] - array[x + 1];
		return (!den) ? std::make_pair(x, array[x])
		              : std::make_pair(x + delta / (2 * den),
		                    array[x] - delta * delta / (8 * den));
	}
	return std::make_pair(x_adjusted, array[x_adjusted]);
}

template std::pair<double, double>
util::parabolic_interpolation<double>(const std::vector<double> &array, int x);
template std::pair<float, float>
util::parabolic_interpolation<float>(const std::vector<float> &array, int x);
